### ISSUE TITLE
(Question) Using Pushy notification listener to increase badge number

### ISSUE THREAD

**Issue by alvarofelipe12** [pushy-me/pushy-cordova#13]
Hello.
I've been developing a push notifications for an Ionic 3 app in which i want to increase the badge number via badge plugin from native resources (@ionic-native/badge) when i receive a push notification. 

While the app is in foreground works perfectly and increase the badge number correctly, but when i try to do the same while the app is in background or closed the listener apparently doesn't works properly because it receive the notification and show it in the notifications of iOS/Android system but it doesn't execute the code in the listener increasing the badge.

```
Pushy.setNotificationListener(function (data) {
          // Print notification payload data
          console.log('Received notification: ' + JSON.stringify(data));
          that.presentAlertNotification(data.message);
          // Increase badge
          that.badge.increase(1);
      });
```

So in conclusion my question is: Is it possible to execute that code inside the listener to increase the badge when the push arrives and the app is closed or in background?

PD: I'm using the version 1.0.21 of pushy-cordova and 1.0.0 of pushy-cordova-receiver.

**Comment by pushy** [comment]
Hi @alvarofelipe12,
I'd be glad to help with your inquiry.

**For iOS**, setting the badge number can simply be done by specifying the `badge` parameter in our [Send Notifications API](https://pushy.me/docs/api/send-notifications) as follows:

```
{
    "to": "a6345d0278adc55d3474f5",
    "data": {
        "message": "Hello World!"
    },
    "notification": {
        "body": "Hello World \u270c",
        "badge": 1,
        "sound": "ping.aiff"
    }
}
```

**For Android**, indeed, the notification listener you set up is not invoked when your app is in the background. This is a Cordova platform restriction. However, it's very easy to customize the behavior of notifications received in the background on Android, by modifying Java push receiver that we ship with the Pushy Cordova SDK which handles incoming notifications.

The built-in push receiver emits a system notification with your app name and icon, and extracts the message from the {"message": "..."} JSON data payload value. 

All you have to do is modify the `PushReceiver.java`, adding a line to set the badge number, using a native Android library like ShortcutBadger:
https://github.com/leolin310148/ShortcutBadger

Modify the `PushReceiver.java` file in the following path:
```
plugins/pushy-cordova/receiver/src/android/PushReceiver.java
```

**Note:** Do not accidentally modify `plugins/pushy-cordova-receiver/src/android/PushReceiver.java` or `platforms/android/src/me/pushy/sdk/cordova/PushReceiver.java` as your modifications will be lost when you reinstall the plugin.

When you are done modifying the `PushReceiver.java` implementation, reinstall the plugin locally for changes to take effect by running the following commands in your root project directory:

```
cordova plugin rm pushy-cordova-receiver
cordova plugin add ./plugins/pushy-cordova/receiver/
````

Please let me know if you were able to get this working.

**Comment by alvarofelipe12** [comment]
Hello and Thanks for the fast response.
I applied to my API the field badge and works properly but i still have a doubt... why the code inside the listener can't be reached while the app is closed in iOS?

Also i noted that in iOS while the app is closed you can receive the push notifications but if i start the app normally it never enters to the listener (pushy.setNotificationListener). But if i open the app by tapping the notification the listener enter and run his code. Why is this happening? Is this a bug? or it's my code running wrong?

And finally I want to clear the badge, is this possible to clear the badge number with pushy?

PS: I'm using now the latest version of pushy-cordova (1.0.25)

**Comment by pushy** [comment]
Hi @alvarofelipe12,
I'd be glad to clarify.

On iOS, your listener will not be invoked when your app is in the background. It will only be invoked when the user interacts with your notification (by clicking on it), or if the app is in foreground, it will be called immediately. This is indeed quite different than on Android. and it is an Apple design decision. 

If you'd like your push notification listener to be invoked when your app is in the background, consider sending "Silent Notifications". 

By sending the `{content_available: true}` flag to our [Send Notifications API](https://pushy.me/docs/api/send-notifications), your listener will be invoked even when your app is in the background. You may also choose not to send any notification styling parameters for silent code execution.

> **Note:** This requires the Background Modes -> **Remote Notifications** capability to be enabled for your iOS project.

Detailed guide:
https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/pushing_updates_to_your_app_silently
